lib/src/js/utils.dart

```diff
@@ -197,6 +200,25 @@ Uri jsToDartUrl(JSUrl url) => Uri.parse(url.toString());
 /// Converts a Dart [Uri] object to a standard JS `URL` object.
 JSUrl dartToJSUrl(Uri url) => JSUrl(url.toString());
 
+/// The use of  `@JSExport()` together with `createDartExport` allow us
+/// to create a JSObject that lazily access containingUrl
+@JSExport()
+class _JSExportCanonicalizeContext {
```

Comment:
This should be in its own file, probably in a `lib/src/js/importer` directory. Also, I'm not sure `@JSExport()` is better than doing the same thing we do for value types and editing the class structure so that we expose JS-friendly members on an existing Dart type. At the very least, this will involve more allocations, which are nice to avoid when possible.

Reply:
I've split the file out. Conceptually I understand what you're saying here, but I'm having a hard time dealing with js interop. I'm aware that this `@JSExport()` will create additional delegate objects, and I settled with it only because it's relatively easy to understand.

@nex3 I'd like to get some help from you on this if you have a way to achieve the same without the delegate.